add --report flag to conftest verify

lanasalameh1 opened the pull request:
add --report flag to conftest verify to get a report and summary of the failing tests for better debugging of failing tests.
more about the flag output and difference from --trace is explained (including screenshots) in github issue associated with the PR
Signed-off-by: lanasalameh1 <lanas@zenity.io>

lanasalameh1 commented:
related issue: https://github.com/open-policy-agent/conftest/issues/576

jalseth commented on internal/commands/verify.go:
```diff
@@ -78,14 +78,22 @@ func NewVerifyCommand(ctx context.Context) *cobra.Command {
-				return fmt.Errorf("output results: %w", err)
+
+			// report currently available with stdout only
+			if runner.Report && runner.Output == output.OutputStandard {
```
As this is, it silently fails to create a report if the user specifies a different output format. I think it would be a better user experience if we check that the output is not supported with the report flag and then surface that error to the user. This also allows us to return earlier, before running any of the tests and removes the need for this nested if/else statement.

lanasalameh1 replied:
agree. fixed.

jalseth commented on output/json.go:
```diff
@@ -45,3 +47,7 @@ func (j *JSON) Output(results []CheckResult) error {
 }
+
+func (j *JSON) Report(results []*tester.Result) error {
+	return nil
```
nit: Since this output isn't supported and because this path shouldn't actually be hit, I think it would be better to return an error here.

jalseth commented on output/junit.go:
```diff
@@ -93,3 +94,7 @@ func getTestName(fileName string, namespace string, message string) string {
 }
+
+func (j *JUnit) Report(results []*tester.Result) error {
+	return nil
```
same as the other

jalseth commented on output/tap.go:
```diff
@@ -85,3 +87,7 @@ func (t *TAP) Output(checkResults []CheckResult) error {
 }
+
+func (t *TAP) Report(results []*tester.Result) error {
+	return nil
```
same as the other

jalseth commented on output/standard.go:
```diff
@@ -159,3 +163,37 @@ func (s *Standard) outputTrace(results []CheckResult, colorizer aurora.Aurora) {
+		Output:      os.Stdout,
+		FailureLine: true}
+
+	dup := make(chan *tester.Result)
+
+	go func() {
```
I don't see the purpose of using a channel here, since the goroutine just iterates over the results so everything is still processed serially. I think this should be removed.

lanasalameh1 replied:
agree. fixed it.

jalseth reviewed (changes requested):
Hi @lanasalameh1 ! Thank you for this PR, it's a really nice feature. I've left some feedback, let me know what you think. It would also be good to add an acceptance test to this before merging.

lanasalameh1 commented:
@jalseth - Thank you for your review.
I addressed your comments, there is one left not resolved - please take a look.
Also, added tests to acceptance.bats - please confirm this is the right place for these tests.

jalseth reviewed (approved):
@lanasalameh1 Apologies for the delay in getting back to you. I've reviewed the changes and they LGTM. Please squash your changes into a single commit with a clear description of the change and I can merge the PR.

lanasalameh1 commented:
Hi James,
All commits are now squashed into a single commit.
Looking forward to merging the PR.

Thanks,
Lana.

On Fri, Jul 23, 2021 at 2:02 AM James Alseth ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> @lanasalameh1 <https://github.com/lanasalameh1> Apologies for the delay
> in getting back to you. I've reviewed the changes and they LGTM. Please
> squash your changes into a single commit with a clear description of the
> change and I can merge the PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-policy-agent/conftest/pull/578#pullrequestreview-713298635>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUGYW6VVIGL4VZEKDKPETPTTZCPPDANCNFSM46F37LFQ>
> .
>

jalseth commented:
@lanasalameh1 Please remove the extra commit messages (`fix tests`, etc.) from the commit message. This way the Git history in the main branch after merging is clear and succinct.

jalseth commented:
Thanks again for the PR!
